### PR TITLE
Rework privatisation of index variables in sequential loops

### PR DESCRIPTION
**Pull request by kiranchandramohan**
When privatising index variables of sequential loops inside an OpenMP parallel region hoist the allocas to the entry block of the parallel region. An interface for Outlineable OpenMP Ops is created to help this. Any OpenMP operation/construct which is eventually outlined should implement this interface. Any allocas for temps or privatised variables should be hoisted to the entry block of such operations.

### CODE REVIEW

File: flang/include/flang/Lower/FIRBuilder.h [hunk 1]
```diff
@@ -69,8 +70,20 @@ class FirOpBuilder : public mlir::OpBuilder {
...
-  /// Get the entry block of the current Function
-  mlir::Block *getEntryBlock() { return &getFunction().front(); }
+  /// Get the block for adding Allocas. If OpenMP is enabled then get the
+  /// the alloca block from an Operation which can be Outlined. Otherwise
+  /// use the entry block of the current Function
+  mlir::Block *getAllocaBlock() {
+    auto iface =
+        getRegion().getParentOfType<mlir::omp::OutlineableOpenMPOpInterface>();
+    return iface ? iface.getAllocaBlock() : &getFunction().front();
+    // mlir::Block *eb = nullptr;
```
Comment (clementval) [inline flang/include/flang/Lower/FIRBuilder.h:80]: Should it be removed?

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+#ifndef OpenMP_OPS_INTERFACES
+#define OpenMP_OPS_INTERFACES
+
+include "mlir/IR/OpBase.td"
+
+def OutlineableOpenMPOpInterface : OpInterface<"OutlineableOpenMPOpInterface"> {
+  let description = [{
+    OpenMP operations whose region will be outlined will implement this
+    interface. These operations will
```
Comment (clementval) [inline mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td:21]: Looks like some words are missing here.

File: flang/include/flang/Lower/FIRBuilder.h [hunk 3]
```diff
@@ -69,8 +70,20 @@ class FirOpBuilder : public mlir::OpBuilder {
   mlir::Value convertWithSemantics(mlir::Location loc, mlir::Type toTy,
                                    mlir::Value val);
 
-  /// Get the entry block of the current Function
-  mlir::Block *getEntryBlock() { return &getFunction().front(); }
```
Comment (schweitzpgi) [inline flang/include/flang/Lower/FIRBuilder.h:73]: We might want to leave `getEntryBlock()` as it reads clearer than `.front()` to me.

File: flang/include/flang/Lower/FIRBuilder.h [hunk 4]
```diff
@@ -69,8 +70,20 @@ class FirOpBuilder : public mlir::OpBuilder {
...
+  /// Get the block for adding Allocas. If OpenMP is enabled then get the
+  /// the alloca block from an Operation which can be Outlined. Otherwise
+  /// use the entry block of the current Function
+  mlir::Block *getAllocaBlock() {
```
Comment (schweitzpgi) [inline flang/include/flang/Lower/FIRBuilder.h:76]: This is growing some header file dependencies and complexity. Can we move it to the `FIRBuilder.cpp` file?

File: flang/test/Lower/OpenMP/omp-seqloop.f90 [hunk 5]
```diff
@@ -21,3 +21,72 @@ program wsloop
...
+    arr(i) = i
+  end do
+!FIRDialect:     omp.terminator
+!FIRDialect:   }
+  !$OMP END PARALLEL
+end subroutine
+
+!FIRDialect: func @_QPsub2() {
+subroutine sub2
+!FIRDialect:   %1 = fir.alloca i32 {bindc_name = "i", uniq_name = "_QFsub2Ei"}
```
Comment (schweitzpgi) [inline flang/test/Lower/OpenMP/omp-seqloop.f90:43]: We typically try to avoid matching ssa-values like `%1` since it doesn't seem robust.

**Review by SouraVX — COMMENTED** [review]
Thanks @kiranchandramohan for the patch :)
I like your approach of having `OutlineableOpenMPOpInterface` that Dialect operations may choose to inherit/implement based on their needs.
I think this is an independent piece of idea/code that should be reviewed in it's own entirety on Phabricator, doing that we will also enable `MLIR` folks to have a look.

File: mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td (hunk 2 above)
Comment (kiranchandramohan) [reply mlir/include/mlir/Dialect/OpenMP/OpenMPOpsInterfaces.td]: Removing the incomplete sentence for now. Don't remember what I was planning to write.

File: flang/include/flang/Lower/FIRBuilder.h (hunk 1 above)
Comment (kiranchandramohan) [reply flang/include/flang/Lower/FIRBuilder.h]: Done.

File: flang/include/flang/Lower/FIRBuilder.h (hunk 3 above)
Comment (kiranchandramohan) [reply flang/include/flang/Lower/FIRBuilder.h:73]: Done.

File: flang/include/flang/Lower/FIRBuilder.h (hunk 4 above)
Comment (kiranchandramohan) [reply flang/include/flang/Lower/FIRBuilder.h]: Done.

File: flang/test/Lower/OpenMP/omp-seqloop.f90 (hunk 5 above)
Comment (kiranchandramohan) [reply flang/test/Lower/OpenMP/omp-seqloop.f90]: Done.

**Comment by SouraVX** [comment]
> > I think this is an independent piece of idea/code that should be reviewed in it's own entirety on Phabricator, doing that we will also enable `MLIR` folks to have a look.
> 
> Plan was to get this reviewed upstream but was not able to think of a way to test.

There were unittests back then but removed after community directions. But I think there's still a hope for at least the reviewal/intent part, we can add a link to this PR for test & more context. Interested folks can have a look.
In this way even in the worst case, community don't want to merge, we can still communicate the overall idea and get thier feedback :)
If community decides against we can always merge here, **no issues** from my side :)

**Review by schweitzpgi — APPROVED** [review]
LGTM

**Pull request merged**